src/Native/Unix/System.Native/pal_io.cpp

```diff
@@ -364,6 +365,60 @@ extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferS
     return 0;
 }
 
+#else
+// To reduce the number of string copies, this function calling pattern works as follows:
+// 1) The managed code calls GetDirentSize() to get the platform-specific
+//    size of the dirent struct.
+// 2) The managed code creates a byte[] buffer of the size of the native dirent
+//    and passes a pointer to this buffer to this function.
+// 3) This function gets a pointer to the possibly statically allocated directory entry.
+// 4) Then, byte[] entry is copied into the byte[] buffer for bufferSize bytes.
+//    This makes the 1st strcpy.
+// 5) The ConvertDirent function will set a pointer to the start of the inode name
+//    in the byte[] buffer so the managed code can find it and copy it out of the
+//    buffer into a managed string that the caller of the framework can use, making
+//    the 2nd and final strcpy.
+//
+//    To make this function thread safe ensure calls on the same DIR* dir never happen concurrently.
+extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferSize, DirectoryEntry* outputEntry)
+{
+    assert(buffer != nullptr);
+    assert(dir != nullptr);
+    assert(outputEntry != nullptr);
+
+    if (bufferSize < static_cast<int32_t>(sizeof(dirent)))
+    {
+        assert(false && "Buffer size too small; use GetDirentSize to get required buffer size");
+        return ERANGE;
+    }
+
+    errno = 0;
+    // Returns a pointer to memory that may be statically allocated by glibc.
+    // Data returned by readdir may be overwritten by other readdir calls 
+    // on the same directory stream.
+    dirent* entry = readdir(dir);
+
+    // positive error number returned -> failure
+    if (errno != 0)
+    {
+        assert(errno == EBADF); // Invalid directory stream descriptor dir.
+        *outputEntry = {}; // managed out param must be initialized
+        return errno;
+    }
+
+    // 0 returned with null result -> end-of-stream
+    if (entry == nullptr)
+    {
+        *outputEntry = {}; // managed out param must be initialized
+        return -1;         // shim convention for end-of-stream
+    }
+
+    memcpy(buffer,entry,static_cast<size_t>(bufferSize));
```

Comment:
Hmm.  I took a look at the managed code.  I was wrong about it not needing the buffer to be populated... or at least, partially wrong.  The entry contains a pointer to a string, so that data needs to be available after this function returns, which means it needs to be in the buffer.  However, even with this copy, it looks like this code is wrong, in two ways:
1) When the memcpy is used to copy from entry to buffer, that will copy the data, but it won't update the pointer value in the entry, so outputEntry will end up pointing to the wrong buffer.
2) This is copying bufferSize bytes, but bufferSize could be larger than entry.

So... some issues that need to be fixed.